File: companies/StackAdapt/576.out-of-boundary-paths.java
import java.util.*;

/**
 * Algorithm:
 * Memoized DFS counts paths from (row, col, moves). Stepping outside the grid
 * contributes one path; running out of moves inside contributes zero.
 *
 * Java data structures:
 * A 3D Integer array stores nullable memoized values, matching Python's
 * lru_cache behavior for the bounded state space.
 *
 * Complexity:
 * Time O(m * n * maxMove), space O(m * n * maxMove).
 */
class Solution {
    private static final int MOD = 1_000_000_007;
    private int m;
    private int n;
    private Integer[][][] memo;

    public int findPaths(int m, int n, int maxMove, int startRow, int startColumn) {
        this.m = m;
        this.n = n;
        this.memo = new Integer[m][n][maxMove + 1];
        return dp(startRow, startColumn, maxMove);
    }

    private int dp(int r, int c, int moves) {
        if (r < 0 || r >= m || c < 0 || c >= n) {
            return 1;
        }
        if (moves == 0) {
            return 0;
        }
        if (memo[r][c][moves] != null) {
            return memo[r][c][moves];
        }
        long total = 0;
        total += dp(r + 1, c, moves - 1);
        total += dp(r - 1, c, moves - 1);
        total += dp(r, c + 1, moves - 1);
        total += dp(r, c - 1, moves - 1);
        memo[r][c][moves] = (int) (total % MOD);
        return memo[r][c][moves];
    }
}

File: Amazon/576.out-of-boundary-paths.cpp
#include <algorithm>
#include <array>
#include <cmath>
#include <climits>
#include <cstdlib>
#include <functional>
#include <numeric>
#include <queue>
#include <random>
#include <regex>
#include <set>
#include <sstream>
#include <string>
#include <tuple>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>
using namespace std;

class Solution {
    static constexpr int MOD = 1000000007;
    int rows, cols;
    vector<vector<vector<int>>> memo;

    int dp(int r, int c, int moves) {
        if (r < 0 || r >= rows || c < 0 || c >= cols) return 1;
        if (moves == 0) return 0;
        int& res = memo[r][c][moves];
        if (res != -1) return res;
        long long total = 0;
        total += dp(r + 1, c, moves - 1);
        total += dp(r - 1, c, moves - 1);
        total += dp(r, c + 1, moves - 1);
        total += dp(r, c - 1, moves - 1);
        return res = total % MOD;
    }

public:
    int findPaths(int m, int n, int maxMove, int startRow, int startColumn) {
        rows = m; cols = n;
        memo.assign(m, vector<vector<int>>(n, vector<int>(maxMove + 1, -1)));
        return dp(startRow, startColumn, maxMove);
    }
};

/*
Interview explanation:
dp(r,c,moves) counts ways to leave the grid from a cell with moves remaining. Moving outside contributes 1; staying inside with zero moves contributes 0.

C++ data structures: a 3D vector memo table stores computed states; int references avoid repeated indexing.

Edge cases: modulo is applied at every memoized state.

Complexity: O(m*n*maxMove) time and space.
*/
